finat/morley.py

```diff
@@ -10,37 +11,66 @@ class Morley(PhysicallyMappedElement, ScalarFiatElement):
+        super().__init__(FIAT.Morley(cell, degree=degree))
 
     def basis_transformation(self, coordinate_mapping):
-        # Jacobians at edge midpoints
-        J = coordinate_mapping.jacobian_at([1/3, 1/3])
-
-        rns = coordinate_mapping.reference_normals()
-        pns = coordinate_mapping.physical_normals()
+        sd = self.cell.get_spatial_dimension()
+        top = self.cell.get_topology()
+        # Jacobians at barycenter
+        bary, = self.cell.make_points(sd, 0, sd+1)
+        J = coordinate_mapping.jacobian_at(bary)
+        detJ = coordinate_mapping.detJ_at(bary)
+        V = identity(self.space_dimension())
 
-        pts = coordinate_mapping.physical_tangents()
+        offset = len(top[sd-2])
+        if sd == 2:
+            pel = coordinate_mapping.physical_edge_lengths()
+            pts = coordinate_mapping.physical_tangents()
+            pns = coordinate_mapping.physical_normals()
+            for e in top[sd-1]:
+                s = offset + e
+                t = partial_indexed(pts, (e,))
+                n = partial_indexed(pns, (e,))
+                nhat = self.cell.compute_normal(e)
+                Jn = J @ Literal(nhat)
+                Bnn = Jn @ n
+                Bnt = Jn @ t
+                V[s, s] = Bnn
+                v = list(top[sd-1][e])
+                V[s, v] = Bnt / pel[e]
+                V[s, v[0]] *= -1
 
-        pel = coordinate_mapping.physical_edge_lengths()
+        else:
+            adjugate = lambda A: ListTensor([[A[1, 1], -1*A[1, 0]], [-1*A[0, 1], A[0, 0]]])
+            edges = self.cell.get_connectivity()[(sd-1, sd-2)]
+            for face in top[sd-1]:
+                s = offset + face
+                thats = self.cell.compute_tangents(sd-1, face)
+                nhat = numpy.cross(*thats)
+                ahat = numpy.linalg.norm(nhat)
+                nhat /= numpy.dot(nhat, nhat)
 
-        V = identity(self.space_dimension())
+                Jn = J @ Literal(nhat)
+                Jt = J @ Literal(thats.T)
+                Gnt = Jn.T @ Jt
+                Gtt = Jt.T @ Jt
+                detG = Gtt[0, 0]*Gtt[1, 1] - Gtt[0, 1]*Gtt[1, 0]
+                area = Power(detG, Literal(0.5))
 
-        for i in range(3):
-            V[i+3, i+3] = (rns[i, 0]*(pns[i, 0]*J[0, 0] + pns[i, 1]*J[1, 0])
-                           + rns[i, 1]*(pns[i, 0]*J[0, 1] + pns[i, 1]*J[1, 1]))
-
-        for i, c in enumerate([(1, 2), (0, 2), (0, 1)]):
-            B12 = (rns[i, 0]*(pts[i, 0]*J[0, 0] + pts[i, 1]*J[1, 0])
-                   + rns[i, 1]*(pts[i, 0]*J[0, 1] + pts[i, 1]*J[1, 1]))
-            V[3+i, c[0]] = -1*B12 / pel[i]
-            V[3+i, c[1]] = B12 / pel[i]
+                Bnn = detJ / area
+                Bnt = Gnt @ adjugate(Gtt) / detG
+                Bnn *= ahat
+                Bnt *= ahat
+                V[s, s] = Bnn
+                V[s, list(edges[face])] = (-1*(Bnt[0] + Bnt[1]), Bnt[0], Bnt[1])
 
         # diagonal post-scaling to patch up conditioning
```

Comment:
Claim: Going to integral moments as suggested above changes the pullback rules slightly but obviates the need for post-scaling.  But if this works now maybe we can do a later clean-up once we understand the transformation theory better.